---
title: "mitoAMOVA_2"
author: "John Wares"
date: "2025-09-04"
output: html_document
---

Here I'm trying to clean up my exploration from last week into cleanest ways to get pairwise divergence metrics and AMOVA for our mtDNA data. First of all, reading in the data and grabbing likely-useful libraries:

```{r getstarted, echo = FALSE}
library(pegas)
library(ape)
library(poppr)
library(adegenet)
library(spider)
library(mmod)

#john
setwd("~/Dropbox/Anderson Geukensia 2025")

#anderson
setwd("C:/Users/andersonsmith/Desktop/R/GitHub/geuk_mito")

```

Next reading in the data, and first trying out a new package called 'haplotypes'. This one I messed around with a few times, at first doing what I thought to be 'random' pulls of sequences from the alignment but because the alignment was organized by the resultant phylogeny (!) it actually threw me for a loop at first until I figured out instead to really look at our data by actual site, that is drawing from the alignment "alphaGeuk3" and using a substring to sort into populations:

```{r haplotypes}

library(haplotypes)

#geuk<-read.fas("shortGeuk.fasta")
#fakepops<-factor(c(rep("A",26),rep("B",26),rep("C",26),rep("D",26),rep("E",26),rep("F",26),rep("G",26),rep("H",26),rep("I",26),rep("J",27)))
#subgeuk<-read.fas("subGeuk.fasta")
#subpops<-factor(c(rep("A",4),rep("B",4),rep("C",4),rep("D",5)))
#GeukAMOVA<-pairPhiST(geuk,fakepops,nperm=100,negatives=FALSE)

#subsubgeuk<-read.fas("geuk100.fasta")
#subsubpops<-factor(c(rep("A",25),rep("B",25),rep("C",25),rep("D",25)))
#GeukAMOVA<-pairPhiST(subsubgeuk,subsubpops,nperm=100,negatives=FALSE,indels="missing")

alphageuk<-read.fas("geuk_sep9.fasta")
locations<-substr(alphageuk@seqnames,1,4)
truepops<-locations

GeukAMOVA<-pairPhiST(alphageuk,truepops,nperm=100,negatives=FALSE,indels="missing") 
GeukAMOVA
phist_mat <- as.dist(GeukAMOVA$PhiST)


```

This will indeed give you pairwise PhiST among sites, and in this case they are spatially organized. Great! However, remember this isn't all the data in the alignment being pulled (I have that now), and I don't seem to be able to get an actual AMOVA from this package. However this package will also calculate pairwise distances, can rearrange haplotypes by different grouping variables, can do *parsimony* gene trees (unusual, and OMG they are trying to revive Templeton's method), put pie charts on a parsimony network, perhaps a few other things. So then there are methods I pulled from NESCENT:

#Nescent AMOVAs
```{r nescent}

library("pegas")
library(apex)
GeukData <- read.multiFASTA(c("geuk_sep9.fasta")) #note changed to the sequences ordered by location now
#probably do not need the read.multiFASTA at this point it is just getting data into genind and doing mostly in POPPR (AMOVA) and pairwise_Gst using (mmod).

#Prepping genind 
plot(GeukData)
locations<-substr(alphageuk@seqnames,1,4)
truepops<-locations
Geuk2<-multidna2genind(GeukData)
Geuk2
pairwise_Gst_Nei(Geuk2)

dists<-dist.multidna(GeukData,pool=TRUE) #pool=TRUE ends up being important so this is a numeric thing

#Splitting the sites up into region models
#NOVA SCOTIA -- now including halifax!
scotia_strata<-data.frame(regions=factor(c(rep("north",21), rep("south",432))), populations=truepops)
strata(Geuk2)<-scotia_strata
setPop(Geuk2)<- ~populations

pegas::amova(dists ~ regions/populations, data=strata(Geuk2),nperm=100)
#highly significant outputs: phict = 0.21, phist = 0.29

#CAPE COD
capecod_strata<-data.frame(regions=factor(c(rep("north",42), rep("south",411))), populations=truepops)
strata(Geuk2)<-capecod_strata
setPop(Geuk2)<- ~populations

pegas::amova(dists ~ regions/populations, data=strata(Geuk2),nperm=100)
#HIGHLY significant again, phict = 0.087, phist = 0.18

#CAPE HATTERAS
hatteras_strata<-data.frame(regions=factor(c(rep("north",171), rep("south",282))), populations=truepops)
strata(Geuk2)<-hatteras_strata
setPop(Geuk2)<- ~populations

pegas::amova(dists ~ regions/populations, data=strata(Geuk2),nperm=100)
#HIGHLY significant again, phict = 0.1027, phist = 0.165

#devtools::install_github('ericarcher/strataG', build_vignettes=TRUE) #argh trying this https://github.com/EricArcher/strataG I did all but step 5 so far and then stymied

#library(strataG)

```

So, this works fine to get pairwise Gst (basically should be pi_tot-pi_subs // pi_tot) and it appears strongly correlated with the PhiST from the other 'haplotypes' output. Remember that these X_st methods give numbers that can be large relative to the actual genomic divergence (net nucleotide divergence), which is what I was trying to do with that package 'strataG' at the end of the chunk but it requires some tinkering with gfortran in OSX and I didn't get all the way there yet. We also now have an AMOVA, which by the way can use 2 different computational flavors:

#poppr AMOVAs
```{r poppr}
#picking up from the NESCENT flavor....

poppr.amova(Geuk2, ~regions/populations, method="ade4") #or "pegas"

```

Other methods I used (including asking a damned LLM for a recipe, grrrr) did not work. Some are deprecated like 'popgenome' and many, circularly, come back to basically the 2 key implementations in either ade4 or pegas. Yet, pegas does a lot but does not seem to do pairwise sample metrics like PhiST, same with 'ade4'. Package 'adegenet' similar limitations. 

#Nucleotide diversities
```{r}
library(pegas)
library(ape)
#need data in DNAbin, it just says x is matrix or list

binGeuk<-as.DNAbin(doneGuek)

mmmpie<-nuc.div(binGeuk,variance=FALSE,pairwise.deletion=TRUE) #doneGuek is not right structure still not 'haplotype' or 'DNAbin'
#ok but this isn't by sample this is across all grrr get so tired of all the data formats
#doneGuek comes from AllGeukSep4_donev2.fasta this is unsatisfactory but works:

Antigo<-binGeuk[1:11,]
Halif<-binGeuk[12:21,]
Maine<-binGeuk[22:28,]
GDPIE<-binGeuk[29:42,]
WooHo<-binGeuk[43:75,]
Goldstr<-binGeuk[76:95,]
Green<-binGeuk[96:114,]
Rutg<-binGeuk[115:128,]
Bivalv<-binGeuk[129:135,]
Maryld<-binGeuk[136:153,]
Virgin<-binGeuk[154:171,]
Roank<-binGeuk[172:178,]
Duke<-binGeuk[179:197,]
Wilmn<-binGeuk[198:207,]
Socar<-binGeuk[208:226,]
Savan<-binGeuk[227:254,]
Sapelo<-binGeuk[255:368,]
Jekyll<-binGeuk[369:399,]
August<-binGeuk[400:412,]
Ponce<-binGeuk[413:437,]
Worth<-binGeuk[438:453,]

TableOne<-c(nuc.div(Antigo,variance=FALSE,pairwise.deletion=TRUE),nuc.div(Halif,variance=FALSE,pairwise.deletion=TRUE),nuc.div(Maine,variance=FALSE,pairwise.deletion=TRUE),nuc.div(GDPIE,variance=FALSE,pairwise.deletion=TRUE),nuc.div(WooHo,variance=FALSE,pairwise.deletion=TRUE),nuc.div(Goldstr,variance=FALSE,pairwise.deletion=TRUE),nuc.div(Green,variance=FALSE,pairwise.deletion=TRUE),nuc.div(Rutg,variance=FALSE,pairwise.deletion=TRUE),nuc.div(Bivalv,variance=FALSE,pairwise.deletion=TRUE),nuc.div(Maryld,variance=FALSE,pairwise.deletion=TRUE),nuc.div(Virgin,variance=FALSE,pairwise.deletion=TRUE),nuc.div(Roank,variance=FALSE,pairwise.deletion=TRUE),nuc.div(Duke,variance=FALSE,pairwise.deletion=TRUE),nuc.div(Wilmn,variance=FALSE,pairwise.deletion=TRUE),nuc.div(Socar,variance=FALSE,pairwise.deletion=TRUE),nuc.div(Savan,variance=FALSE,pairwise.deletion=TRUE),nuc.div(Sapelo,variance=FALSE,pairwise.deletion=TRUE),nuc.div(Jekyll,variance=FALSE,pairwise.deletion=TRUE),nuc.div(August,variance=FALSE,pairwise.deletion=TRUE),nuc.div(Ponce,variance=FALSE,pairwise.deletion=TRUE),nuc.div(Worth,variance=FALSE,pairwise.deletion=TRUE))
```

#Mantel test
```{r}
#Load package (sorry John!!!)
#install.packages("geosphere")
library(geosphere)
library(ade4)


#Load latlongs in as csv, north-south
latlongs_geuk <- read_csv("./latlongs.csv")
#Generate Haversine distance matrix for latlongs
haversine_distm <- distm(latlongs_geuk) %>% 
  as.dist()

#Use phist matrix, run Mantel in ade4
mantel(haversine_distm, GeukAMOVA, method = "pearson", permutations = 99)


```

#Map graphic
```{r}
#Install and load packages
#install.packages("mapdata")
#install.packages("maptools", 
                 #repos = "https://packagemanager.posit.co/cran/2023-10-13")
library (tidyverse)
library(mapdata)

#Make df of all sampling points
geuk_lat <- c(45.6679, 44.0142447, 42.759, 41.576, 40.8968, 40.3700, 39.31141, 
              38.0228, 37.306988, 32.76158333, 32.038, 26.7171667,
              29.134, 29.833, 31.392, 35.903, 34.702, 34.175, 44.65069)
geuk_lon <- c(-61.8783, -69.5375049, -70.891, -70.639, -73.4341, -73.3014, 
              -74.2, -75.2144, -76.411197, -79.96671667, -81.048, -80.078,
              -80.955, -80.955, -81.271, -75.604, -76.751, -77.932, -63.42344)

geuk_dat <- data.frame(lat=geuk_lat, lon=geuk_lon)

#Map of US, Canada, and Mexico with sample sites included
usa <- map_data("usa")
canada <- map_data("worldHires", "Canada")
mexico <- map_data("worldHires", "Mexico")

NAmap <- ggplot() + geom_polygon(data = usa, 
                                 aes(x=long, y = lat, group = group), 
                                 fill = "white", 
                                 color = "black") +
  geom_polygon(data = canada, aes(x=long, y = lat, group = group), 
               fill = "white", color="black") + 
  geom_polygon(data = mexico, aes(x=long, y = lat, group = group), 
               fill = "white", color="black") +
  coord_fixed(xlim = c(-81, -60),  ylim = c(25, 48), ratio = 1.1) +
  geom_point(data = geuk_dat,
             aes(x = lon,
            y = lat), 
            color = "mediumorchid3",
            size = 3.3)+
  annotate("text", y = 35.5, x = -74, label = "A", size = 7) +
  annotate("text", y = 41.5, x = -69, label = "B", size = 7) +
   annotate("text", y = 42.5, x = -65.5, label = "C", size = 7) +
  xlab("Longitude") +
  ylab("Latitude") +
  theme_minimal()

NAmap 

ggsave("gss_musselmap.pdf",
       plot = NAmap,
       device = "pdf",
       scale = 1,
       dpi = "print",
       limitsize = TRUE)
```
